File: bin/tacticaloffsite.py
import subprocess
import argparse
import datetime as dt

__version__ = "1.0.1"

OFFSITE_DEST = "tactical@offsite.tacticaltechs.com"


def main():
    parser = argparse.ArgumentParser(description="Tactical Offsite Backup")
    parser.add_argument(
        "--source", action="store", dest="source", type=str, required=False
    )
    parser.add_argument(
        "--dest", action="store", dest="destination", type=str, required=False
    )
    parser.add_argument(
        "--limit", action="store", dest="limit", type=int, required=False
    )
    parser.add_argument(
        "--logfile",
        action="store",
        dest="logfile",
        type=str,
        required=False,
        help="The full path to the log file",
    )
    parser.add_argument(
        "--version", required=False, action="store_true", help="Prints version"
    )
    args = parser.parse_args()

    if args.version:
        print(__version__)
        return

    # sync the metadata file first
    cmd1 = [
        "rsync",
        "-havz",
        "--progress",
        f"--bwlimit={args.limit}",
        "--partial-dir=.rsync-partial",
        "--include='*.vbm'",
        "--exclude='*'",
        args.source,
        f"{OFFSITE_DEST}:{args.destination}",
    ]
    # next, sync the full backups
    cmd2 = [
        "rsync",
        "-havz",
        "--progress",
        f"--bwlimit={args.limit}",
        "--partial-dir=.rsync-partial",
        "--include='*.vbk'",
        "--exclude='*'",
        args.source,
        f"{OFFSITE_DEST}:{args.destination}",
    ]
    # finally, sync the incremental backups
    cmd3 = [
        "rsync",
        "-havz",
        "--progress",
        f"--bwlimit={args.limit}",
        "--partial-dir=.rsync-partial",
        "--include='*.vib'",
        "--exclude='*'",
        args.source,
        f"{OFFSITE_DEST}:{args.destination}",
    ]

    started = dt.datetime.now().strftime("%b-%d-%Y_%I-%M-%S_%p")
    with open(args.logfile, "a+") as f:
        f.write(f"Offsite started at {started}\n")

    with open(args.logfile, "a+") as f:

        r1 = subprocess.run(cmd1, stdout=f, stderr=subprocess.STDOUT)
        r2 = subprocess.run(cmd2, stdout=f, stderr=subprocess.STDOUT)
        r3 = subprocess.run(cmd3, stdout=f, stderr=subprocess.STDOUT)

    finished = dt.datetime.now().strftime("%b-%d-%Y_%I-%M-%S_%p")
    with open(args.logfile, "a+") as f:
        f.write(f"\nOffsite finished at {finished}\n")
        f.write("-" * 100)
        f.write("\n")


if __name__ == "__main__":
    main()

File: bin/linux_agent/tacticaloffsite.go
// go build -ldflags="-X 'main.RsyncHost=sshuser@example.com'" tacticaloffsite.go
package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"os/exec"
	"strings"
)

var (
	Version   = "1.1.0"
	RsyncHost string
)

func doRsync(source string, dest string, limit int, logfile *os.File, ext string) (err error) {

	rsyncArgs := []string{
		"-havz", "--progress", fmt.Sprintf("--bwlimit=%d", limit),
		"--partial-dir=.rsync-partial", fmt.Sprintf("--include='*.%s'", ext),
		"--exclude='*'", source, fmt.Sprintf("%s:%s", RsyncHost, dest),
	}

	cmd := exec.Command("rsync", rsyncArgs...)
	cmd.Stdout = logfile
	cmd.Stderr = logfile

	err = cmd.Start()
	if err != nil {
		fmt.Println(err)
	}

	err = cmd.Wait()
	if err != nil {
		fmt.Println(err)
	}

	return nil

}

func main() {

	source := flag.String("source", "", "Source Dir")
	dest := flag.String("dest", "", "Destination Dir")
	limit := flag.Int("limit", 300, "bw limit in kb/s")
	logfile := flag.String("logfile", "", "Log file location")
	ver := flag.Bool("version", false, "Prints version")
	flag.Parse()

	if *ver {
		fmt.Println(Version)
		return
	}

	if len(os.Args) != 9 {
		fmt.Printf("Usage: %s source dest limit logfile\r\n", os.Args[0])
		os.Exit(1)
	}

	f, err := os.OpenFile(*logfile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		log.Fatal(err)
		os.Exit(1)
	}
	defer f.Close()

	logger := log.New(f, "", log.LstdFlags)
	logger.Println("Offsite Started")

	// metadata, full backups, incremental backups
	veeams := []string{"vbm", "vbk", "vib"}

	for _, val := range veeams {
		err = doRsync(*source, *dest, *limit, f, val)
		if err != nil {
			log.Fatal(err)
			os.Exit(1)
		}
	}

	logger.Println("Offsite Finished")
	f.WriteString(fmt.Sprintf("%s\n", strings.Repeat("-", 100)))
}
